### ISSUE TITLE
[Bug]: PC0038 Terminating AL methods like FieldError not treated like Error(...) in control-flow analysis (PC0038 false positive; likely affects other rules too)

### ISSUE THREAD

**Issue by lv-janpieter** [ALCops/Analyzers#463]
### Analyzer

PlatformCop

### Issue type

False positive (diagnostic fires when it shouldn't)

### Rule ID

PC0038

### Description

ALCops appears to treat only the built-in `Error(...)` as a hard procedure terminator. The idiomatic AL alternative `Record.FieldError(...)` — which also never returns — is not recognized as terminating. That causes false positives wherever control flow depends on “this call ends the procedure.”

Confirmed for **PC0038** (*Not all code paths return a value*). The same gap likely affects other **ALCops** rules that model scope-leaving statements.

**Expected behavior**

Any ALCops analysis that treats `Error(...)` as ending a path should also treat `FieldError(...)` the same way.

For PC0038 specifically: no diagnostic when every branch either returns a value or ends in `Error(...)` or `FieldError(...)`.

**Actual behavior**

PC0038 is reported when the only non-returning path ends with `FieldError`, even though that branch cannot continue at runtime.

Replacing `FieldError` with `Error(...)` removes the diagnostic, although both have the same control-flow effect.

The [PC0038 documentation](https://alcops.dev/docs/analyzers/platformcop/pc0038/) explicitly accepts paths that end with `Error(...)`. `FieldError` should be treated the same way.

**Suggested fix**

1. Introduce (or extend) a shared helper for **unconditional terminating statements** used by control-flow analysis.
2. Include at least: `Error(...)`, `FieldError(...)` (both overloads).
3. Audit ALCops rules that currently special-case only `Error(...)` (PC0038, complexity, FC0007 scope-leaving, etc.).

### AL code to reproduce

Minimal reproduction:

```al
codeunit 50100 "PC0038 FieldError Repro"
{
    local procedure GetDescriptionFromType(SalesLine: Record "Sales Line"): Text[100]
    begin
        case SalesLine.Type of
            SalesLine.Type::Item:
                exit(SalesLine.Description);

            SalesLine.Type::"G/L Account":
                exit(SalesLine.Description);

            else
                SalesLine.FieldError(Type); // PC0038 false positive
        end;
    end;

    local procedure GetDescriptionFromType(SalesLine: Record "Sales Line"): Text[100]
    begin
        if SalesLine.Type = SalesLine.Type::Item then
            exit(SalesLine.Description)
        else
            SalesLine.FieldError(Type); // PC0038 false positive
    end;
}
```

Observed: PC0038 on GetDescriptionFromType.

Workaround that clears the diagnostic (same control flow intent):

```al

codeunit 50100 "PC0038 FieldError Repro"
{
    local procedure GetDescriptionFromType(SalesLine: Record "Sales Line"): Text[100]
    begin
        case SalesLine.Type of
            SalesLine.Type::Item:
                exit(SalesLine.Description);
            SalesLine.Type::"G/L Account":
                exit(SalesLine.Description);
            else
                Error('%1 must not be %2.', SalesLine.FieldCaption(Type), SalesLine.Type);
        end;
    end;
}
```
Observed: PC0038 is not reported.

**Why this matters**
FieldError is the idiomatic AL way to raise a field-specific validation error (correct field caption / value in the message). Forcing developers to replace it with Error(...) only to satisfy PC0038 either:

weakens the error message, or leads to unnecessary suppressions / dummy exit('') after FieldError.

**Suggested fix**
Treat FieldError as terminating statements in the PC0038 control-flow analysis, consistent with Error(...).

### Broader concern: shared “terminator” model (ALCops scope)

This looks less like a one-off PC0038 bug and more like a **shared incompleteness in how ALCops models scope-leaving / terminating statements**.

Today, ALCops seems to special-case:

- `Error(...)`
- (and often) `exit` / `exit(...)`

…but not `Record.FieldError(...)`, even though it is the standard AL idiom for a field-specific validation error that always aborts execution.

Other **ALCops** diagnostics that may need the same terminator set:

1. **PC0038** — missing return on a path (confirmed false positive with `FieldError`)
2. **Cognitive / cyclomatic complexity** (e.g. LC0089 / LC0090 / LC0009 / LC0010) — if early-exit / guard scoring treats `Error(...)` specially but ignores `FieldError`, scores can be inconsistent for equivalent guard clauses
3. **Formatting / scope-leaving rules** (e.g. FC0007) — docs describe blank-line rules for scope-leaving statements as `exit` and built-in `Error(...)`; `FieldError` is the same kind of statement from a reader’s perspective and may be handled inconsistently

**Ask:** Please centralize a list of known **unconditional** terminating AL methods and reuse it across ALCops analyzers, rather than hardcoding only `Error(...)` per rule.

For this issue, the only additional candidate beyond `Error(...)` is Record.FieldError(...). We deliberately do **not** propose adding methods like `TestField`, because those are not unconditionally terminating.

### Why this matters

`FieldError` is the idiomatic AL way to raise a field-specific validation error. Forcing `Error(...)` only to satisfy analyzers either weakens the message or pushes dummy `exit('')` / suppressions. Recognizing `FieldError` as a terminator in ALCops benefits multiple rules and keeps diagnostics consistent with how AL actually runs.

### Related Microsoft issue (AA0136 — separate report, out of scope for ALCops)

Microsoft CodeCop **AA0136** (*Do not write code that will never be hit*) shows the inverse asymmetry: it recognizes `Error(...)` as terminating for unreachable-code detection, but **not** `FieldError`. That cannot be fixed in ALCops and would need a separate report on [microsoft/AL](https://github.com/microsoft/AL).

```al
local procedure DemoUnreachable(SalesLine: Record "Sales Line")
begin
    SalesLine.FieldError(Type);
    Message('Unreachable?'); // Not flagged by AA0136

    Error('Stop');
    Message('Unreachable?'); // Is flagged by AA0136
end;
```

**Comment by Arthurvdv** [comment]
## Investigation findings (branch `release/v1.1.0`)

Confirmed. We audited every ALCops rule that special-cases `Error(...)` as a terminating / scope-leaving statement, and cross-checked the terminator set against the NAV SDK sources (v18.0 symbol tables). Findings below are meant as the working brief for the implementing agent.

### Affected rules (3)

**1. PC0038 — root cause confirmed**
`src/ALCops.PlatformCop/Analyzers/NotAllCodePathsReturnValue.cs`
The flow analysis models a terminated path as an empty state set, and its private `IsFlowTerminatingCall` recognizes exactly two names on `MethodKind.BuiltInMethod`: `Error` and `ThrowError`. `FieldError` is absent, which produces the reported false positive. Two additional observations:
- The name comparison uses `StringComparison.Ordinal`, unlike the rest of the codebase which uses `SemanticFacts.IsSameName` / `SemanticFacts.NameEqualityComparer` (AL names are case-insensitive). Worth aligning in the same change.
- `ThrowError` matches no AL built-in: the SDK's `ErrorInfoClassTypeSymbol` defines no `Throw`/`ThrowError` member (the `DataError.ThrowError` string in `CodeGenFacts` is a codegen-time C# runtime helper, not AL surface). The entry is effectively dead.

**2. LC0089 / LC0089i / LC0090 (cognitive complexity) — confirmed, plus a structural gap**
`src/ALCops.LinterCop/Analyzers/CognitiveComplexity.cs`
`guardClauseExitCommands` is `{Break, Continue, Error, Quit, Skip}`, and `FieldError` is missing. But adding the name alone is not enough: the guard-command check only accepts a member-access receiver when it is `CurrReport`/`CurrXMLport` (via `guardClauseIdentifiers`), so `Rec.FieldError(...)` is structurally rejected regardless of the name set. Detection here is purely syntactic (no semantic model / `MethodKind` check), unlike PC0038 and FC0007 — the implementing agent should account for that asymmetry.

**3. FC0007 (scope-leaving blank line, opt-in) — confirmed**
`src/ALCops.FormattingCop/Analyzers/StatementBlocksSeparatedByBlankLine.cs`
`IsBuiltInErrorInvocation` matches only `Error` (correctly via `IsSameName` + `MethodKind.BuiltInMethod`). `FieldError` is the same kind of scope-leaving statement from the reader's perspective and is currently not covered. The rule's design doc (`.github/instructions/fc0007-statement-blocks-separated-by-blank-line.instructions.md`) already flags the drift risk between this rule and LC0090's guard heuristic.

### Rules checked and NOT affected

- **LC0009 / LC0010** (cyclomatic complexity / maintainability index, `src/ALCops.LinterCop/Analyzers/CyclomaticComplexityAndMaintainabilityIndex.cs`): pure syntax-kind counting; no guard-clause discount and no `Error` special-casing exists today, so there is nothing for `FieldError` to be inconsistent with.
- **PC0034, LC0048**: reference `Error` only for message/placeholder content checks, not control flow.
- **PC0030, PC0027, LC0097**: model `exit` statements only; no built-in terminator concept.

### No shared terminator helper exists

Each of the three affected rules carries its own private predicate — exactly the duplication/drift the issue calls out. There is currently no shared classification of flow-terminating built-ins in `src/ALCops.Common`. For reference, the repo's established pattern for a single source of truth of built-in method name sets is `src/ALCops.Common/RecordMethodClassification.cs` (governed by `.github/instructions/common-library.instructions.md`, which admits utilities needed by two or more cop projects — satisfied here with three consumers). We deliberately leave the concrete design to the implementing agent.

### Terminator set, verified against the NAV SDK

The defensible set of **unconditionally** terminating AL built-ins is exactly:

| Method | Container | Overloads |
|---|---|---|
| `Error` | `Dialog` (global) | `Error(Text[, params])`, `Error(ErrorInfo)` |
| `FieldError` | `Table` (any `Record`) | `FieldError(Field[, Text])`, `FieldError(Field, ErrorInfo)` |
| `FieldError` | `FieldRef` | `FieldError([Text])`, `FieldError(ErrorInfo)` |

Supporting evidence from the SDK sources:
- The SDK's class documentation resources describe only `Error` ("ends the execution of AL code") and `FieldError` ("stops the execution of the code") with terminal phrasing — out of ~4300 built-in doc entries.
- There is no `[DoesNotReturn]`-style flag or method-ID enum in the SDK; built-ins can only be identified by name + `MethodKind.BuiltInMethod` (+ optionally the containing class symbol). All relevant containers (`Dialog`, `Table`, `FieldRef`) derive from `LanguageClassTypeSymbol`, so the symbol shape is identical for `Error` and `FieldError`.
- The SDK itself has no control-flow/reachability infrastructure to reuse — ALCops has to own this classification.

Deliberately **excluded** (agreeing with the reporter on `TestField`):
- `TestField(...)` — conditional, errors only on failure.
- `CurrReport.Break/Quit/Skip`, `CurrXMLport.Break/Quit/Skip` — trigger-scoped with deferred runtime effect; they stay handled where they are today (LC0089/LC0090 guard identifiers) and should not enter a general "never returns" set.
- `Debugger.Stop()` — controls a debugged session, not the calling code.
- `ErrorInfo.ThrowError` — does not exist in the AL surface (see PC0038 note above).

### Acceptance criteria for the fix

**PC0038** (`src/ALCops.PlatformCop.Test/Rules/NotAllCodePathsReturnValue/`):
- No diagnostic for both repro shapes from this issue (`case ... else Rec.FieldError(Type);` and `if/else` with `FieldError` in the else branch), for named and unnamed returns.
- Cover both `Record.FieldError` overloads and `FieldRef.FieldError`.
- A **user-defined** procedure named `FieldError` (or `Error`) must still be treated as non-terminating (the `MethodKind.BuiltInMethod` guard covers this — keep a test proving it).
- Existing `Error`-terminator fixtures to clone: `NoDiagnostic/UnnamedIfElseErrorTerminates.al`, `NamedIfElseErrorTerminates.al`, `UnnamedCaseElseErrorTerminates.al`, `UnnamedGuardClauseErrorFirst.al`. Note: current PC0038 fixtures contain no table object, so `FieldError` fixtures need a table (or `FieldRef`) declaration.
- Keep `.github/instructions/pc0038-not-all-code-paths-return-value.instructions.md` (design-decision table + test inventory) in sync.

**LC0089/LC0090** (`src/ALCops.LinterCop.Test/Rules/CognitiveComplexity/`):
- `if <cond> then Rec.FieldError(...)` receives the same guard-clause discount as `if <cond> then Error(...)`; complexity scores for the two equivalent guard shapes must be equal.
- Clone from `NoDiagnostic/IfStatementGuardClause.al` / `CurrReportGuardClause.al` fixtures.

**FC0007** (`src/ALCops.FormattingCop.Test/Rules/StatementBlocksSeparatedByBlankLine/`):
- Under `ScopeLeavingMode` including Error, `Rec.FieldError(...)` requires the same preceding blank line as `Error(...)` (Has/NoDiagnostic pair). Tests need the rule's `.ruleset.json` and `alcops.json` via `MemoryFileSystem` (rule is opt-in/disabled by default).
- Update `.github/instructions/fc0007-statement-blocks-separated-by-blank-line.instructions.md` accordingly.

**Cross-cutting:**
- Whatever shared classification is introduced must use `SemanticFacts.NameEqualityComparer` / `IsSameName` (case-insensitive), and remain symbol-based where a semantic model is available (`MethodKind.BuiltInMethod`) so user-defined `Error`/`FieldError` procedures never match.
- Docs on alcops.dev for PC0038 (and FC0007/LC0089 if their docs enumerate scope-leaving statements) should mention `FieldError` alongside `Error`.

### AA0136

Agreed this is out of scope for ALCops and needs a separate report on microsoft/AL. One data point worth including there: AA0136's implementation guard (`ContainingSymbol.Kind == SymbolKind.Class`) already passes for `FieldError` — only the literal name comparison against `"error"` excludes it.

---
Generated with Claude Fable 5

**Comment by MODUSCarstenScholling** [comment]
@Arthurvdv I am going to handle this.
